From f7ce3ecb24e67e12f3e8b4f5fc4c21ab2849b887 Mon Sep 17 00:00:00 2001
From: Mingliang Hu <mhu4@marvell.com>
Date: Fri, 26 Mar 2010 11:20:46 +0800
Subject: [PATCH] SD card: Fix a bug to compute the clock of MM4CLK pin.

Signed-off-by: Lou Mao <ylmao@marvell.com>

When the clock needed is a little smaller than the base clock. The variable 'shift' will become a negative.
---
 drivers/mmc/pxa9xx_sdh.c |   14 +++++++++++---
 1 files changed, 11 insertions(+), 3 deletions(-)

diff --git a/drivers/mmc/pxa9xx_sdh.c b/drivers/mmc/pxa9xx_sdh.c
index 285f58d..051ad96 100644
--- a/drivers/mmc/pxa9xx_sdh.c
+++ b/drivers/mmc/pxa9xx_sdh.c
@@ -271,9 +271,17 @@ void pxasdh_set_ios(struct mmc *mmc)
 		shift = 0;
 		while (clk >= (0x1 << shift))
 			shift++;
-		clkrt = (0x1 << (shift - 2));
-		if ((host->clk_rate / (clkrt * 2)) > mmc->clock)
-			clkrt *= 2;
+		if(shift == 1)
+			clkrt = 1;
+		else if (shift < 1){
+			printf("host clock is lower than the clock needed, it should be a bug.\n");
+			while(1);
+		}
+		else {
+			clkrt = (0x1 << (shift - 2));
+			if ((host->clk_rate / (clkrt * 2)) > mmc->clock)
+				clkrt *= 2;
+		}
 	}
 	sdh_set_reg(mmc, MMC4_CNTL2,
 		sdh_get_reg(mmc, MMC4_CNTL2) & ~MMC4CLKEN);
-- 
1.6.0.4

